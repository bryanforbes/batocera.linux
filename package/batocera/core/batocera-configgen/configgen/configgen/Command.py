from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
