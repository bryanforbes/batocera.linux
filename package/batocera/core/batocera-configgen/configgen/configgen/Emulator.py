from __future__ import annotations

import collections.abc
import logging
import xml.etree.ElementTree as ET
from dataclasses import InitVar, dataclass, field
from typing import TYPE_CHECKING, Any, ClassVar, Literal, TypeVar, overload

import yaml

from .batoceraPaths import BATOCERA_CONF, BATOCERA_SHADERS, DEFAULTS_DIR, ES_SETTINGS, USER_SHADERS
from .settings.unixSettings import UnixSettings
from .utils.missing import MISSING, MissingType

if TYPE_CHECKING:
    from pathlib import Path
    from typing_extensions import deprecated

    from .gun import GunMapping

_logger = logging.getLogger(__name__)
_T = TypeVar('_T')
_TV = TypeVar('_TV')
_FV = TypeVar('_FV')


# adapted from https://gist.github.com/angstwad/bf22d1822c38a92ec0a9
def _dict_merge(destination: dict[str, Any], source: collections.abc.Mapping[str, Any]) -> None:
    """ Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
    updating only top-level keys, dict_merge recurses down into dicts nested
    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
    ``dct``.
    :param destination: dict onto which the merge is executed
    :param source: dict merged into destination
    :return: None
    """
    for key, value in source.items():
        if (key in destination and isinstance(destination[key], dict) and isinstance(value, collections.abc.Mapping)):
            _dict_merge(destination[key], value)
        else:
            destination[key] = value


@dataclass(slots=True)
class Emulator:
    TRUE_VALUES: ClassVar[set[Literal['1', 'true', 'on', 'enabled', True]]] = {'1', 'true', 'on', 'enabled', True}
    MISSING: ClassVar = MISSING

    name: str
    rom: InitVar[Path]

    config: dict[str, Any] = field(init=False)
    renderconfig: dict[str, Any] = field(init=False)

    def __post_init__(self, rom: Path, /) -> None:
        # read the configuration from the system name
        self.config = self.__get_system_config(
            DEFAULTS_DIR / "configgen-defaults.yml",
            DEFAULTS_DIR / "configgen-defaults-arch.yml"
        )
        if "emulator" not in self.config or not self.config["emulator"]:
            _logger.error("no emulator defined. exiting.")
            raise Exception("No emulator found")

        gsname = self.game_settings_name(rom)

        # load configuration from batocera.conf
        batoceraSettings = UnixSettings(BATOCERA_CONF)
        globalSettings = batoceraSettings.load_all('global')
        controllersSettings = batoceraSettings.load_all('controllers', includeName=True)
        systemSettings = batoceraSettings.load_all(self.name)
        folderSettings = batoceraSettings.load_all(f'{self.name}.folder["{rom.parent}"]')
        gameSettings = batoceraSettings.load_all(f'{self.name}["{gsname}"]')

        # add some other options
        displaySettings = batoceraSettings.load_all('display')
        for opt, value in displaySettings.items():
            self.config[f"display.{opt}"] = value

        # update config
        Emulator.update_configuration(self.config, controllersSettings)
        Emulator.update_configuration(self.config, globalSettings)
        Emulator.update_configuration(self.config, systemSettings)
        Emulator.update_configuration(self.config, folderSettings)
        Emulator.update_configuration(self.config, gameSettings)
        self.updateFromESSettings()
        _logger.debug("uimode: %s", self.config['uimode'])

        # forced emulators ?
        self.config["emulator-forced"] = False
        self.config["core-forced"] = False
        if "emulator" in globalSettings or "emulator" in systemSettings or "emulator" in gameSettings:
            self.config["emulator-forced"] = True
        if "core" in globalSettings or "core" in systemSettings or "core" in gameSettings:
            self.config["core-forced"] = True

        # update renderconfig
        self.renderconfig = {}
        if "shaderset" in self.config:
            if self.config["shaderset"] != "none":
                rendering_defaults = USER_SHADERS / "configs" / str(self.config["shaderset"]) / "rendering-defaults.yml"
                if rendering_defaults.exists():
                    self.renderconfig = self.__get_yaml_defaults(
                        rendering_defaults,
                        rendering_defaults.with_name("rendering-defaults-arch.yml")
                    )
                else:
                    rendering_defaults = BATOCERA_SHADERS / "configs" / str(self.config["shaderset"]) / "rendering-defaults.yml"
                    self.renderconfig = self.__get_yaml_defaults(
                        rendering_defaults,
                        rendering_defaults.with_name("rendering-defaults-arch.yml")
                    )
            elif self.config["shaderset"] == "none":
                rendering_defaults = BATOCERA_SHADERS / "configs" / "rendering-defaults.yml"
                self.renderconfig = self.__get_yaml_defaults(
                    rendering_defaults,
                    rendering_defaults.with_name("rendering-defaults-arch.yml")
                )

        # for compatibility with earlier Batocera versions, let's keep -renderer
        # but it should be reviewed when we refactor configgen (to Python3?)
        # so that we can fetch them from system.shader without -renderer
        systemSettings = batoceraSettings.load_all(f'{self.name}-renderer')
        gameSettings = batoceraSettings.load_all(f'{self.name}["{gsname}"]-renderer')

        # es only allow to update systemSettings and gameSettings in fact for the moment
        Emulator.update_configuration(self.renderconfig, systemSettings)
        Emulator.update_configuration(self.renderconfig, gameSettings)

    @property
    def core(self) -> str:
        return self.config['core']

    @property
    def core_forced(self) -> bool:
        return self.config['core-forced']

    @property
    def emulator(self) -> str:
        return self.config['emulator']

    @property
    def emulator_forced(self) -> bool:
        return self.config['emulator-forced']

    @property
    def ui_mode(self) -> Literal['Full', 'Kiosk', 'Kid']:
        return self.config['uimode']

    @property
    def show_fps(self) -> bool:
        return self.config['showFPS']

    def game_settings_name(self, rom: Path) -> str:
        # sanitize rule by EmulationStation
        # see FileData::getConfigurationName() on batocera-emulationstation
        gs_name = rom.name.replace('=','')
        gs_name = gs_name.replace('#','')
        _logger.info("game settings name: %s", gs_name)
        return gs_name

    def __get_yaml_defaults(self, default_yaml: Path, default_arch_yaml: Path, /) -> dict[str, Any]:
        with default_yaml.open() as f:
            defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader)

        arch_defaults: dict[str, Any] = {}
        if default_arch_yaml.exists():
            with default_arch_yaml.open() as f:
                loaded_arch_defaults = yaml.load(f, Loader=yaml.CLoader)

            if loaded_arch_defaults is not None:
                arch_defaults = loaded_arch_defaults

        config: dict[str, Any] = {}

        if 'default' in defaults:
            config = defaults['default']

        if 'default' in arch_defaults:
            _dict_merge(config, arch_defaults['default'])

        if self.name in defaults:
            _dict_merge(config, defaults[self.name])

        if self.name in arch_defaults:
            _dict_merge(config, arch_defaults[self.name])

        return config

    def __get_system_config(self, default_yaml: Path, default_arch_yaml: Path, /) -> dict[str, Any]:
        defaults = self.__get_yaml_defaults(default_yaml, default_arch_yaml)

        result: dict[str, Any] = { 'emulator': defaults['emulator'], 'core': defaults['core'] }

        # In the yaml files, the "options" structure is not flat, so we have to flatten it here
        # because the options are flat in batocera.conf to make it easier for end users to edit
        if 'options' in defaults:
            _dict_merge(result, defaults['options'])

        return result

    def has_option(self, key: str, /) -> bool:
        return key in self.config

    @overload
    def get_option(self, key: str, default: MissingType = ..., /) -> Any | MissingType:
        ...

    @overload
    def get_option(self, key: str, default: _T, /) -> Any | _T:
        ...

    def get_option(self, key: str, default: _T | MissingType = MISSING, /) -> Any | _T | MissingType:
        return self.config.get(key, default)

    @overload
    def get_option_bool(self, key: str, default: bool = False, /, *, return_values: None = None) -> bool:
        ...

    @overload
    def get_option_bool(self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV]) -> _TV | _FV:
        ...

    def get_option_bool(self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV] | None = None) -> bool | _TV | _FV:
        value = self.config.get(key, MISSING)

        if value is MISSING:
            return default

        if isinstance(value, str):
            value = value.lower()

        if return_values is None:
            return value in self.TRUE_VALUES
        else:
            return return_values[value not in self.TRUE_VALUES]

    @overload
    def get_option_str(self, key: str, default: MissingType = ..., /) -> str | MissingType:
        ...

    @overload
    def get_option_str(self, key: str, default: str, /) -> str:
        ...

    def get_option_str(self, key: str, default: str | MissingType = MISSING, /) -> str | MissingType:
        value = self.config.get(key, MISSING)

        if value is MISSING:
            return default

        return str(value)

    @overload
    def get_option_int(self, key: str, default: MissingType = ..., /) -> int | MissingType:
        ...

    @overload
    def get_option_int(self, key: str, default: int, /) -> int:
        ...

    def get_option_int(self, key: str, default: int | MissingType = MISSING, /) -> int | MissingType:
        value = self.config.get(key, MISSING)

        if value is MISSING:
            return default

        return int(value)

    @overload
    def get_option_float(self, key: str, default: MissingType = ..., /) -> float | MissingType:
        ...

    @overload
    def get_option_float(self, key: str, default: float, /) -> float:
        ...

    def get_option_float(self, key: str, default: float | MissingType = MISSING, /) -> float | MissingType:
        value = self.config.get(key, MISSING)

        if value is MISSING:
            return default

        return float(value)

    def option_items(self, /, *, starts_with: str | None = None) -> collections.abc.Iterator[tuple[str, Any]]:
        if starts_with is None:
            yield from self.config.items()
        else:
            starts_with_len = len(starts_with)
            for key, value in self.config.items():
                if key.startswith(starts_with):
                    yield key[starts_with_len:], value


    if TYPE_CHECKING:
        @deprecated('Use has_option()')
        def isOptSet(self, key: str) -> bool:
            ...

        @deprecated('Use get_option_bool()')
        def getOptBoolean(self, key: str) -> bool:
            ...

        @deprecated('Use get_option_str()')
        def getOptString(self, key: str) -> str:
            ...
    else:
        def isOptSet(self, key: str) -> bool:
            return key in self.config

        def getOptBoolean(self, key: str) -> bool:
            value = self.config.get(key)

            if isinstance(value, str):
                value = value.lower()

            return value in self.TRUE_VALUES

        def getOptString(self, key: str) -> str:
            if key in self.config:  # noqa: SIM102
                if self.config[key]:
                    return self.config[key]
            return ""

    def get_guns_borders_size(self, guns: GunMapping, /) -> str | None:
        borders_size = self.get_option_str("controllers.guns.borderssize", "medium")

        # overriden by specific options
        borders_mode = "normal"
        if (config_borders_mode := self.get_option_str("controllers.guns.bordersmode", "auto")) != "auto":
            borders_mode = config_borders_mode
        if (config_borders_mode := self.get_option_str("bordersmode", "auto")) != "auto":
            borders_mode = config_borders_mode

        # others are gameonly and normal
        if borders_mode == "hidden":
            return None

        if borders_mode == "force":
            return borders_size

        for gun in guns.values():
            if gun.need_borders:
                return borders_size

        return None

    def get_guns_borders_ratio(self, guns: GunMapping, /) -> str | None:
        # returns None to follow the bezel overlay size by default
        return self.get_option("controllers.guns.bordersratio", None)

    @staticmethod
    def update_configuration(config: dict[str, Any], settings: dict[str, Any], /) -> None:
        # ignore all values "default", "auto", "" to take the system value instead
        # ideally, such value must not be in the configuration file
        # but historically some users have them

        config.update({
            key: value for key, value in settings.items()
            if value and value != 'default' and value != 'auto'
        })

    # fps value is from es
    def updateFromESSettings(self):
        try:
            esConfig = ET.parse(ES_SETTINGS)

            # showFPS
            drawframerate_node = esConfig.find("./bool[@name='DrawFramerate']")
            drawframerate_value = drawframerate_node.attrib["value"] if drawframerate_node is not None else 'false'
            if drawframerate_value not in ['false', 'true']:
                drawframerate_value = 'false'

            self.config['showFPS'] = drawframerate_value == 'true'

            # uimode
            uimode_node = esConfig.find("./string[@name='UIMode']")
            uimode_value = uimode_node.attrib["value"] if uimode_node is not None else 'Full'
            if uimode_value not in ['Full', 'Kiosk', 'Kid']:
                uimode_value = 'Full'

            self.config['uimode'] = uimode_value

        except Exception:
            self.config['showFPS'] = False
            self.config['uimode'] = "Full"
