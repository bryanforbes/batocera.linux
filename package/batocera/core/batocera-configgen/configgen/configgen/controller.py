from __future__ import annotations

import xml.etree.ElementTree as ET
from collections.abc import Mapping
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import TYPE_CHECKING, Final, Literal, Self, TypeAlias, TypedDict, Unpack, cast

from .batoceraPaths import BATOCERA_ES_DIR, USER_ES_DIR
from .input import Input, InputDict

if TYPE_CHECKING:
    import argparse

"""Default mapping of Batocera keys to SDL_GAMECONTROLLERCONFIG keys."""
_DEFAULT_SDL_MAPPING: Final = {
    'b': 'a',
    'a': 'b',
    'x': 'y',
    'y': 'x',
    'l2': 'lefttrigger',
    'r2': 'righttrigger',
    'l3': 'leftstick',
    'r3': 'rightstick',
    'pageup': 'leftshoulder',
    'pagedown': 'rightshoulder',
    'start': 'start',
    'select': 'back',
    'up': 'dpup',
    'down': 'dpdown',
    'left': 'dpleft',
    'right': 'dpright',
    'joystick1up': 'lefty',
    'joystick1left': 'leftx',
    'joystick2up': 'righty',
    'joystick2left': 'rightx',
    'hotkey': 'guide'
}


def _key_to_sdl_game_controller_config(keyname: str, name: str, type: str, id: str, value: str) -> str | None:
    """
    Converts a key mapping to the SDL_GAMECONTROLLER format.

    Arguments:
      keyname: (str) SDL_GAMECONTROLLERCONFIG input name.
      name: (str) `es_input.cfg` input name.
      type: (str) 'button', 'hat', or 'axis'
      id: (int) Numeric key id.
      value: (int) Hat value. Only used if type == 'hat' or type == 'axis' and 'joystick' in name.
    Returns:
      (str) SDL_GAMECONTROLLERCONFIG-formatted key mapping string.
    Examples:
      _keyToSdlGameControllerConfig('leftshoulder', 'l1', 'button', 6)
        'leftshoulder:b6'

      _keyToSdlGameControllerConfig('dpleft', 'left', 'hat', 0, 8)
        'dpleft:h0.8'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, -1)
        'lefty:a1'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, 1)
        'lefty:a1~'

      _keyToSdlGameControllerConfig('dpup', 'up', 'axis', 1, -1)
        'dpup:-a1'
    """
    if type == 'button':
        return f'{keyname}:b{id}'
    elif type == 'hat':
        return f'{keyname}:h{id}.{value}'
    elif type == 'axis':
        if 'joystick' in name:
            return f"{keyname}:a{id}{'~' if int(value) > 0 else ''}"
        elif keyname in ('dpup', 'dpdown', 'dpleft', 'dpright'):
            return f"{keyname}:{'-' if int(value) < 0 else '+'}a{id}"
        else:
            return f'{keyname}:a{id}'
    elif type == 'key':
        return None
    else:
        raise ValueError(f'unknown key type: {type!r}')


class _ControllerChanges(TypedDict, total=False):
    guid: str
    player: int | None
    index: int
    real_name: str
    device_path: str | None
    button_count: int | None
    hat_count: int | None
    axis_count: int | None
    physical_device_path: str | None
    physical_id: int | None


@dataclass(slots=True, kw_only=True)
class Controller:
    device_name: str
    type: Literal['keyboard', 'joystick']
    guid: str
    player: int | None
    index: int = -1
    real_name: str = ""
    inputs: InputDict = field(default_factory=dict)
    device_path: str | None = None
    button_count: int | None = None
    hat_count: int | None = None
    axis_count: int | None = None
    physical_device_path: str | None = None
    physical_id: int | None = None

    @property
    def uid_name(self) -> str:
        return self.guid + self.device_name

    def replace(self, /, **changes: Unpack[_ControllerChanges]) -> Self:
        return replace(self, **changes)

    def generate_sdl_game_db_line(self, sdl_mapping: Mapping[str, str] = _DEFAULT_SDL_MAPPING, /) -> str:
        config = [self.guid, self.real_name, 'platform:Linux']

        def add_mapping(input: Input) -> None:
            key_name = sdl_mapping.get(input.name, None)
            if key_name is None:
                return
            sdl_config = _key_to_sdl_game_controller_config(key_name, input.name, input.type, input.id, input.value)
            if sdl_config is not None:
                config.append(sdl_config)

        hotkey_input: Input | None = None
        mapped_button_ids: set[str] = set()
        for input in self.inputs.values():
            if input.name is None:
                continue
            if input.name == 'hotkey':
                hotkey_input = input
                continue
            if input.type == 'button':
                mapped_button_ids.add(input.id)

            add_mapping(input)

        if hotkey_input is not None and hotkey_input.id not in mapped_button_ids:
            add_mapping(hotkey_input)

        config.append('')

        return ','.join(config)

    @classmethod
    def from_element(cls, element: ET.Element, /) -> Self:
        return cls(
            device_name=cast(str, element.get("deviceName")),
            type=cast(Literal['keyboard', 'joystick'], element.get("type")),
            guid=cast(str, element.get("deviceGUID")),
            player=None,
            index=-1,
            inputs=Input.from_parent_element(element)
        )

    @classmethod
    def parse_all(cls) -> list[Self]:
        return [
            cls.from_element(controller)
            for conffile in [BATOCERA_ES_DIR / "es_input.cfg", USER_ES_DIR / 'es_input.cfg'] if conffile.exists()
            for controller in ET.parse(conffile).getroot().findall(".//inputConfig")
        ]

    @classmethod
    def load_all(cls) -> dict[str, Self]:
        return {controller.uid_name: controller for controller in cls.parse_all()}

    @classmethod
    def load_for_players(cls, max_players: int, args: argparse.Namespace, /) -> dict[int, Self]:
        controllers: dict[int, Self] = {}
        all_controllers = cls.load_all()

        for player_number in range(1, max_players + 1):
            controller = cls.find_best_controller_config(all_controllers, args, player_number)
            if controller is not None:
                controllers[player_number] = controller

        return controllers

    @classmethod
    def find_best_controller_config(
        cls, controllers: Mapping[str, Self], args: argparse.Namespace, player_number: int, /,
    ) -> Self | None:
        player_index: int | None = getattr(args, f'p{player_number}index')

        if player_index is None:
            return None

        player_guid: str = getattr(args, f'p{player_number}guid')
        player_name: str = getattr(args, f'p{player_number}name')
        player_device_path: str = getattr(args, f'p{player_number}devicepath')
        player_buttons: int = getattr(args, f'p{player_number}nbbuttons')
        player_hats: int = getattr(args, f'p{player_number}nbhats')
        player_axes: int = getattr(args, f'p{player_number}nbaxes')

        for controller in controllers.values():
            if controller.guid == player_guid and controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.guid == player_guid:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        return None


ControllerMapping: TypeAlias = Mapping[str, Controller]
ControllerDict: TypeAlias = Mapping[str, Controller]
ControllerPlayerMapping: TypeAlias = Mapping[int, Controller]
ControllerPlayerDict: TypeAlias = dict[int, Controller]


def generate_sdl_game_controller_config(controllers: ControllerPlayerMapping, /) -> str:
    return '\n'.join([controller.generate_sdl_game_db_line() for controller in controllers.values()])

def write_sdl_controller_db(
    controllers: ControllerPlayerMapping, output_file: str | Path = '/tmp/gamecontrollerdb.txt', /,
) -> Path:
    output_file = Path(output_file)
    with output_file.open('w') as text_file:
        text_file.write(generate_sdl_game_controller_config(controllers))
    return output_file
