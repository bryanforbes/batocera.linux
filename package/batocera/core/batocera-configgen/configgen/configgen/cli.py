from __future__ import annotations

import logging
import time
from collections.abc import Callable
from functools import partial
from pathlib import Path
from typing import Any, TypedDict, TypeVar, Unpack, cast

import click

from . import profiler
from .emulators import run_emulator
from .utils.logger import setup_logging

_FC = TypeVar('_FC', bound=Callable[..., Any] | click.Command)

_logger = logging.getLogger(__name__)


def _player_callback(player: int, ctx: click.Context, option: click.Option, value: int | None) -> int | None:
    current_value: list[PlayerOptions] = ctx.params.setdefault('players', [])

    option_names = (
        f'p{player}hat_count',
        f'p{player}button_count',
        f'p{player}device_path',
        f'p{player}real_name',
        f'p{player}guid',
        f'p{player}index',
    )

    items = [
        (f'p{player}axis_count', value),
        *[(option_name, ctx.params.pop(option_name)) for option_name in option_names],
    ]

    if all(item[1] is not None for item in items):
        p_length = len(f'p{player}')
        current_value.append(
            cast(
                PlayerOptions,
                {
                    'player_number': player,
                    **{option_name[p_length:]: option_value for option_name, option_value in items},
                },
            )
        )
        return value

    if all(item[1] is None for item in items):
        return value

    raise click.UsageError(f'All options that start with p{player} must be specified if any are specified', ctx=ctx)


def _players(max_players: int) -> Callable[[_FC], _FC]:
    def wrap(callable: _FC) -> _FC:
        for n in range(max_players, 0, -1):
            callable = click.option(f'-p{n}nbaxes', f'p{n}axis_count', type=int, callback=partial(_player_callback, n))(
                callable
            )
            callable = click.option(f'-p{n}nbhats', f'p{n}hat_count', type=int)(callable)
            callable = click.option(f'-p{n}nbbuttons', f'p{n}button_count', type=int)(callable)
            callable = click.option(f'-p{n}devicepath', f'p{n}device_path', type=str)(callable)
            callable = click.option(f'-p{n}name', f'p{n}real_name', type=str)(callable)
            callable = click.option(f'-p{n}guid', type=str)(callable)
            callable = click.option(f'-p{n}index', type=int)(callable)

        return callable

    return wrap


class PlayerOptions(TypedDict):
    player_number: int
    index: int
    guid: str
    real_name: str
    device_path: str
    button_count: int
    hat_count: int
    axis_count: int


class Options(TypedDict):
    system: str
    rom: Path
    emulator: str | None
    core: str | None
    netplaymode: str | None
    netplaypass: str | None
    netplayip: str | None
    netplayport: str | None
    netplaysession: str | None
    state_slot: str | None
    state_filename: Path | None
    autosave: str | None
    systemname: str | None
    gameinfoxml: Path
    lightgun: bool
    wheel: bool
    trackball: bool
    spinner: bool
    players: list[PlayerOptions]


@click.command()
@click.option('-system', required=True)
@click.option('-rom', required=True, type=click.Path(path_type=Path))
@click.option('-emulator')
@click.option('-core')
@click.option('-netplaymode')
@click.option('-netplaypass')
@click.option('-netplayip')
@click.option('-netplayport')
@click.option('-netplaysession')
@click.option('-state_slot')
@click.option('-state_filename', type=click.Path(path_type=Path))
@click.option('-autosave')
@click.option('-systemname')
@click.option('-gameinfoxml', default='/dev/null', type=click.Path(path_type=Path))
@click.option('-lightgun', is_flag=True)
@click.option('-wheel', is_flag=True)
@click.option('-trackball', is_flag=True)
@click.option('-spinner', is_flag=True)
@_players(8)
def cli(**kwargs: Unpack[Options]) -> None:
    with setup_logging():
        exit_code = -1

        try:
            exit_code = run_emulator(kwargs)
        except Exception:
            _logger.exception('configgen exception: ')

        profiler.disable(dump=True)

        time.sleep(1)  # this seems to be required so that the gpu memory is restituated and available for es
        _logger.debug('Exiting configgen with status %s', exit_code)

        exit(exit_code)
