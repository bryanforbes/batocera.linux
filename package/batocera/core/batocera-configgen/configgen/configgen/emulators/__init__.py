from __future__ import annotations

from contextlib import nullcontext
from importlib import import_module
from typing import TYPE_CHECKING, Final

from ..batoceraPaths import BATOCERA_CONF
from ..config import RenderConfig, SystemConfig
from ..controller import Controller
from ..controllersConfig import getGamesMetaData
from ..gun import Gun, precalibrate_guns
from ..settings.unixSettings import UnixSettings
from ..utils.squashfs import squashfs_rom
from ..utils.wheels import configure_wheels
from ._base import Emulator as Emulator  # noqa: TCH001

if TYPE_CHECKING:
    from ..cli import Options

_GENERATOR_MAP: Final[dict[str, str]] = {}


def get_emulator_cls(emulator: str, /) -> type[Emulator]:
    module_path = _GENERATOR_MAP.get(emulator, emulator.replace('-', '_'))

    try:
        module = import_module(f'..{module_path}', package=__name__)
        emulator_cls: type[Emulator] = module.Emulator
    except ImportError as e:
        if e.name is not None and e.name.startswith(__name__.split('.')[0]):
            raise Exception(f'no emulator found for {emulator!r}') from e

        raise
    except AttributeError as e:
        raise Exception(f'no emulator found for {emulator!r}') from e

    return emulator_cls


def run_emulator(options: Options, /) -> int:
    exit_code = -1
    original_rom = options['rom']

    if original_rom.suffix == '.squashfs':
        rom_context_manager = squashfs_rom(original_rom)
    else:
        rom_context_manager = nullcontext(original_rom)

    with rom_context_manager as rom:
        system = options['system']

        settings = UnixSettings(BATOCERA_CONF)
        system_config = SystemConfig.load(options, settings)
        render_config = RenderConfig.load(options, settings, system_config.get('shaderset'))

        metadata = getGamesMetaData(system, original_rom)

        controllers = Controller.load_for_player_options(options['players'])

        if system_config.use_guns:
            guns = Gun.get_all()
            precalibrate_guns(system, system_config.emulator, system_config.core, original_rom)
        else:
            guns = {}

        with configure_wheels(controllers, system, system_config, metadata) as (controllers, wheels):
            emulator_cls = get_emulator_cls(system_config.emulator)
            emulator = emulator_cls(system, system_config, render_config, metadata, controllers, guns, wheels)
            exit_code = emulator.run(rom, original_rom, options['gameinfoxml'])

    return exit_code
