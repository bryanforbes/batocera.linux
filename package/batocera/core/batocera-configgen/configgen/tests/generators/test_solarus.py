from __future__ import annotations

from typing import TYPE_CHECKING

import pytest

from configgen.batoceraPaths import CONFIGS, ROMS
from configgen.generators.solarus.solarusGenerator import SolarusGenerator
from tests.generators.base import GeneratorBaseTest
from tests.mock_controllers import make_player_controller_list

if TYPE_CHECKING:
    from syrupy.assertion import SnapshotAssertion

    from configgen.controller import Controller, Controllers
    from configgen.Emulator import Emulator


@pytest.mark.usefixtures('fs')
class TestSolarusGenerator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[SolarusGenerator]:
        return SolarusGenerator

    @pytest.fixture
    def system_name(self) -> str:
        return 'solarus'

    @pytest.fixture
    def emulator(self) -> str:
        return 'solarus'

    def test_generate(
        self,
        generator: SolarusGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'solarus' / 'rom.solarus',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert (CONFIGS / 'solarus' / 'pads.ini').read_text() == snapshot(name='pads')

    @pytest.mark.parametrize(
        'mock_system_config',
        [
            {'joystick': 'normal'},
            {'joystick': 'joystick1'},
            {'joystick': 'joystick2'},
        ],
        ids=str,
    )
    def test_generate_joystick(
        self,
        generator: SolarusGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'solarus' / 'rom.solarus',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (CONFIGS / 'solarus' / 'pads.ini').read_text() == snapshot(name='pads')

    def test_generate_keyboard(
        self,
        generator: SolarusGenerator,
        mock_system: Emulator,
        keyboard_controller: Controller,
        generic_xbox_pad: Controller,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'solarus' / 'rom.solarus',
            make_player_controller_list(keyboard_controller, generic_xbox_pad),
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (CONFIGS / 'solarus' / 'pads.ini').read_text() == snapshot(name='pads')

    def test_generate_no_controllers(
        self,
        generator: SolarusGenerator,
        mock_system: Emulator,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'solarus' / 'rom.solarus',
            [],
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (CONFIGS / 'solarus' / 'pads.ini').read_text() == snapshot(name='pads')
