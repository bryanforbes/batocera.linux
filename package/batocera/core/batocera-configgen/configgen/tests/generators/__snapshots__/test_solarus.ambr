# serializer version: 1
# name: TestSolarusGenerator.test_generate
  tuple(
    list([
      'solarus-run',
      '-fullscreen=yes',
      '-cursor-visible=no',
      '-lua-console=no',
      '-quit-combo=8+9',
      '-joypad-num1=0',
      '/userdata/roms/solarus/rom.solarus',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
      'SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS': '0',
    }),
  )
# ---
# name: TestSolarusGenerator.test_generate[pads]
  '''
  action=button 0
  attack=button 1
  item1=button 3
  item2=button 2
  pause=button 9
  right=hat 0 right
  up=hat 0 up
  left=hat 0 left
  down=hat 0 down
  
  '''
# ---
# name: TestSolarusGenerator.test_generate_joystick[{'joystick': 'joystick1'}][pads]
  '''
  action=button 0
  attack=button 1
  item1=button 3
  item2=button 2
  pause=button 9
  up=axis 1 -
  down=axis 1 +
  left=axis 0 -
  right=axis 0 +
  
  '''
# ---
# name: TestSolarusGenerator.test_generate_joystick[{'joystick': 'joystick2'}][pads]
  '''
  action=button 0
  attack=button 1
  item1=button 3
  item2=button 2
  pause=button 9
  up=axis 3 -
  down=axis 3 +
  left=axis 2 -
  right=axis 2 +
  
  '''
# ---
# name: TestSolarusGenerator.test_generate_joystick[{'joystick': 'normal'}][pads]
  '''
  action=button 0
  attack=button 1
  item1=button 3
  item2=button 2
  pause=button 9
  right=hat 0 right
  up=hat 0 up
  left=hat 0 left
  down=hat 0 down
  
  '''
# ---
# name: TestSolarusGenerator.test_generate_keyboard[pads]
  '''
  action=None
  attack=None
  item1=None
  item2=None
  pause=None
  right=None
  up=None
  left=None
  down=None
  
  '''
# ---
# name: TestSolarusGenerator.test_generate_no_controllers[pads]
  ''
# ---
# name: TestSolarusGenerator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTALT',
        'KEY_F4',
      ]),
    }),
    'name': 'solarus',
  })
# ---
